### ISSUE TITLE
repeated custom options cannot be parsed

### ISSUE THREAD

**Issue by hugebdu** [dcodeIO/protobuf.js#1142]
protobuf.js version: 6.8.8

[repeated](https://stackoverflow.com/a/28758624/10233273) [custom options](https://developers.google.com/protocol-buffers/docs/proto3#custom_options) are not supported due to being mapped to object type in [`ReflectionObject#options`](https://github.com/dcodeIO/protobuf.js/blob/master/src/object.js#L26-L30 
Maybe should be an array?

**Comment by petebarnett** [comment]
@rclark Just found your post while working with the same set of proto files from Envoy.
It affects all custom [options](https://developers.google.com/protocol-buffers/docs/reference/proto3-spec#option) used inline with a message or enum.
I'm now trying to work around it by pre-processing the files to remove these, since they result in no code generation anyway.
Ideally, I'd love to implement [Lyft protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate) for Protobuf.js, but I don't know enough yet about extension points for parsing or codegen. Not sure if it's even possible.

**Comment by eyalpost** [comment]
@nicolasnoble @alexander-fenster Are there any intentions to fix this?
I'm willing to give it a try and make a PR but want to first check with you about which solution fits best here since this can break existing clients.
Should we add an option that indicates whether to parse repeated options?

**Comment by alexander-fenster** [comment]
@eyalpost We haven't looked at this one yet, so the below text is just some general thoughts.

When we looked at similar issues before, the questions that we asked often were "how does `protoc` behave in this case", and "what's written in the [protobuf Language Guide](https://developers.google.com/protocol-buffers/docs/proto3)". For the latter question, I was unable to find an exact answer now, so, I think, the `protoc` behavior would be our best baseline to look at.

Anyway, the current suggestion here in this issue (replace an option value with a list of values) can break the current users so cannot be released without a semver major release. I wonder if there is a compatible and non-ugly way of making it work without breaking any of the existing code :)

**Comment by hugebdu** [comment]
@alexander-fenster how about leaving `ReflectionObject#options` the way it is today and introducing a new property `ReflectionObject#rawOptions` of type `Array`?
Similar to node's [`http.IncomingMessage#rawHeaders`](https://nodejs.org/api/http.html#http_message_rawheaders)?

**Comment by alexander-fenster** [comment]
@hugebdu That might be a good idea since adding the new property won't break the existing users and those who need repeated custom options can just start using it. Feel free to send a PR and we'll be happy to review it!

**Comment by eyalpost** [comment]
Another question before starting to implement.
So assuming we have the following proto:

```proto
extend google.protobuf.FieldOptions {
    repeated int32 flags = 50000;
}

message TestFieldOptionsInt {
    string field1 = 2 [(flags) = 555, (flags) = 666];
    string field2 = 3 [(flags) = 777];
}
```
I would expect `field1.rawOptions["(flags)"]` to return `[555,666]`
But what would you expect `field2.rawOptions["(flags)"]` to return?
Just `777` ? Or `[777]`?

**Comment by alexander-fenster** [comment]
I personally would expect it to return `[777]`, so that when a new option gets added to the proto (which by itself is a non-breaking change), this change will not break the existing code. I'd love to see other folks' thoughts on that (@murgatroid99, @nicolasnoble WDYT?)

**Comment by eyalpost** [comment]
I tend to agree, but the main problem is that currently the parser does not resolve the option definition so it has no knowledge whether the option is repeated or not.
I suggest we start with `rawOptions` as a mechanism to simply collect the options from the proto definition "as-is" (Implemented in PR #1254 ) and will always appear as an array of options - This means for the above case, the rawOptions for `field1` will be `[{flag: 555}, {flag: 666}]` and for `field2` it will be `[{flag: 777}]` which means the client is responsible for finding *all* the values for a flag that they are interested in.

**Comment by alexander-fenster** [comment]
@eyalpost I think this is fine, and this is exactly the _raw_ options as the name suggests, so the user can choose how they process the options. We'll take a look at the PR early next week! Thank you for handling this.

**Comment by eyalpost** [comment]
There's a limitation with the current implementation with message based options. Would love to hear your feedback on this.
Consider the following scenario:
```proto
message Rule {
  string val = 1;
  string val2 = 2;
}
extend google.protobuf.MessageOptions { 
    repeated Rule rule = 50000; 
}

message TestOptions {
    option (rule) = {val : "X"};
    option (rule) = {val2: "Y", val: "Z"};
}
```
The current implementation will produce these `rawOptions`:
```json
[
 {"(rule).val": "X"},
 {"(rule).val2": "Y"},
 {"(rule).val": "Z"}
]
```
Which means there's no way to reconstruct the original definition.
I already have a POC which produces the following instead:
```json
[
  {"(rule)": {"val": "X"}},
  {"(rule)": {"val2": "Y", "val": "Z"}}
]
```
but I need to give it some more work before it's ready for review

**Comment by eyalpost** [comment]
Opened a new PR: https://github.com/protobufjs/protobuf.js/pull/1256 which implements proper parsing of options

**Comment by eyalpost** [comment]
@alexander-fenster note that this new PR still keeps the current `options` field intact to make sure it doesn’t break existing clients

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
